### ISSUE TITLE
add consecutive=True to the group command

### ISSUE THREAD

**Issue by devlounge** [rethinkdb/rethinkdb#5507]
Hi,

This is a feature request:

example sequence:
[{"foo": 1}], {"foo":1}, {"foo":0}, {"foo": 1}]
sequence.group(index='foo', consecutive=True) => 3 groups instead of 2 currently

The expected behaviour would be the same than itertools.groupby in python

**Comment by thelinuxlich** [comment]
I think you're looking for ```r([{"foo": 1}, {"foo":1}, {"foo":0}, {"foo": 1}]).group(r.row.values())```

**Comment by danielmewes** [comment]
Hi @devlounge ,

the good news is that you will be able to easily do this yourself in RethinkDB 2.3 by using the `fold` command that we are introducing. The details of `fold` are discussed in https://github.com/rethinkdb/rethinkdb/issues/3736#issuecomment-163095798 and following , but here is how you would implement the consecutive group command:

```js
r.expr([{"foo": 1}, {"foo":1}, {"foo":0}, {"foo": 1}])
 .fold({},
         // Accumulator function.
         // Starts a new group or adds an element to the current group's `reduction`
         function(current_group, input_element) {
           // The `branch` is to start a new group when necessary
           return r.branch(
               current_group("group").eq(input_element("foo")).default(false),
               current_group.merge({reduction: current_group("reduction").append(input_element)}),
               {group: input_element("foo"), reduction: [input_element]});
         },
         {
           // Emitter function. Emits the old_group accumulator if we just started a new group.
           emit: function(old_group, input_element, new_group) {
               return r.branch(old_group("group").ne(new_group("group")), [old_group], []).default([]);
           },
           // Final emitter function. Emit the final group if there is any
           finalEmit: function(final_group) {
               return r.branch(final_group.ne({}), [final_group], []);
           }
         })
```

(@chipotle: Would this be a suitable example for the `fold` API page? Maybe for the cookbook?)

As a side remark: This can't actually work when using an index with the `group`, since an indexed group will always traverse the sequence in the order of the index (so the individual groups will be contiguous in any case).

I'll leave this open to discuss adding this as a native feature. However I think it's a lower priority thing since you can already implement it yourself with the above query.

**Comment by danielmewes** [comment]
@thelinuxlich I think that's only a part of it, since he wants a new group to be started whenever the value of "foo" changes. Your command would only have one group per "foo" value in the output, even if the input goes back and forth between multiple values of "foo".

**Comment by thelinuxlich** [comment]
You're right, I wrote that without really testing :(

**Comment by thelinuxlich** [comment]
By the way, does ```fold``` works on streams?

**Comment by danielmewes** [comment]
Yes `fold` can be used on streams and also emits a stream which is one reason for why it's so useful. :-)
It's similar to `concatMap`, except that it keeps an accumulator across multiple input documents so you can maintain state.

**Comment by thelinuxlich** [comment]
Awesome
